File: 198. House Robber.go
package main

import (
	"fmt"
)

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func dynamicProgrammingSolution(index int, nums []int, cache map[int]int) int {
	if index >= len(nums) {
		return 0
	}

	if _, didFind := cache[index]; didFind {
		return cache[index]
	}

	cache[index] = max(dynamicProgrammingSolution(index+1, nums, cache), nums[index]+dynamicProgrammingSolution(index+2, nums, cache))
	return cache[index]
}

func rob(nums []int) int {
	return dynamicProgrammingSolution(0, nums, map[int]int{})
}

func main() {
	ans := rob([]int{1, 2, 1, 1})
	fmt.Println("answer is: ", ans)
}

File: 198. House Robber.py
class Solution:

    def topDownDPRob(self, nums: List[int]) -> int:

        cache = dict()

        def rcr(index):    
            if index in cache: return cache[index]
            elif index >= len(nums): return 0
            cache[index] = max(rcr(index+1),rcr(index+2) + nums[index])
            return cache[index] 
        return rcr(0)

    def iterativeRob(self, nums: List[int]) -> int:
        
        rob1, rob2 = 0, 0
        for n in nums:
            temp = max(rob1+n,rob2)
            rob1 = rob2
            rob2 = temp
        return rob2